internal/controller/ipam_controller.go

```diff
@@ -0,0 +1,204 @@
+		}
+	}
+	return nil
+}
+
+// generateCIDRPool generates a CIDRPool object for the given Rail
+func (r *NvIPAMReconciler) generateCIDRPool(rail *config.Rail, hosts []config.Host, crossRailNet string) *nvipamv1.CIDRPool {
+	staticAllocations := getHostsAllocation(rail.Name, hosts)
+	pool := &nvipamv1.CIDRPool{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      rail.Name,
+			Namespace: r.CIDRPoolsNamespace,
+			Labels: map[string]string{
+				spectrumXIPAMControllerName: "",
+			},
+		},
+		Spec: nvipamv1.CIDRPoolSpec{
+			CIDR:                 rail.Subnet,
+			GatewayIndex:         ptr.To[int32](0),
+			PerNodeNetworkPrefix: 31,
+			StaticAllocations:    staticAllocations,
+			Routes: []nvipamv1.Route{
+				{
+					Dst: rail.Subnet,
+				},
+				{
+					Dst: crossRailNet,
+				},
+			},
+		},
+	}
+	pool.ObjectMeta.ManagedFields = nil
+	pool.SetGroupVersionKind(nvipamv1.GroupVersion.WithKind(cidrPoolType))
+	return pool
+}
+
+// getHostsAllocation  computes the static allocations for the hosts for a specific Rail
+func getHostsAllocation(rail string, hosts []config.Host) []nvipamv1.CIDRPoolStaticAllocation {
+	allocs := make([]nvipamv1.CIDRPoolStaticAllocation, 0)
+	for _, h := range hosts {
+		for _, r := range h.Rails {
+			if r.Name == rail {
+				alloc := nvipamv1.CIDRPoolStaticAllocation{
+					NodeName: h.HostID,
+					Prefix:   r.Network,
+					Gateway:  strings.Split(r.Network, "/")[0],
+				}
+				allocs = append(allocs, alloc)
+			}
+		}
+	}
+	return allocs
+}
+
+// reconcileDelete delete created CIDRPools and remove finalizer
+func (r *NvIPAMReconciler) reconcileDelete(ctx context.Context, cm *corev1.ConfigMap) (ctrl.Result, error) {
+	logr := log.FromContext(ctx)
+	logr.Info("Reconciling delete")
+	if err := r.deleteCIDRPools(ctx); err != nil {
+		return ctrl.Result{}, err
+	}
+	logr.Info("Removing finalizer")
+	controllerutil.RemoveFinalizer(cm, spectrumXFinalizer)
+	cm.ObjectMeta.ManagedFields = nil
+	if err := r.Client.Patch(ctx, cm, client.Apply, client.ForceOwnership, client.FieldOwner(spectrumXIPAMControllerName)); err != nil {
+		return ctrl.Result{}, fmt.Errorf("error while removing finalizer to configmap %w", err)
+	}
+	return ctrl.Result{}, nil
+}
+
+// deleteCIDRPools deletes all the CIDR pools created by controller
+func (r *NvIPAMReconciler) deleteCIDRPools(ctx context.Context) error {
+	p := &unstructured.Unstructured{}
+	p.SetGroupVersionKind(nvipamv1.GroupVersion.WithKind(cidrPoolType))
+	if err := r.Client.DeleteAllOf(ctx, p, client.InNamespace(r.CIDRPoolsNamespace), client.MatchingLabels{
+		spectrumXIPAMControllerName: "",
+	}); err != nil {
+		return fmt.Errorf("error while removing all %s: %w", p.GetObjectKind().GroupVersionKind().String(), err)
+	}
+	return nil
+}
+
+// SetupWithManager sets up the controller with the Manager.
+func (r *NvIPAMReconciler) SetupWithManager(mgr ctrl.Manager) error {
+	configMapPredicate := predicate.NewPredicateFuncs(func(obj client.Object) bool {
```

Comment:
I think currently it is possible to filter this on the manager level https://github.com/Mellanox/network-operator-init-container/blob/ae52bad7fbc110af549bb18d93a8d0056747e015/cmd/network-operator-init-container/app/app.go#L100

Reply:
Done